speed up the ISR by removing float arithmetic, array loop, search loo…

IDC-Dragon opened the pull request:
Excuse me again, with another speed optimization :-). 
The ISR is now free of loops and float/64bit arithmetic, just maintains a summation value and stability criteria + history.  (BTW, I think the check of stable_freq_history didn't do the intended, now fixed.)
The sample rate calculation and validation now happens in the foreground functions, which should be executed less frequent.

IDC-Dragon commented:
Maybe I should re-introduce 64 bit timestamps, but only for the variables `now` and `prev_time`, not necessary for the calculation. That would close a tiny hole I opened in `spdif_rx_get_status()`: Every hour when the 32bit timer wraps, it may report true for 10ms, despite no interrupts.
However, sharing something >32 bits machine size between ISR and foreground is unsafe, too, because access is not atomic any more...

elehobica commented:
Thanks for this.
For speed optimization of checking sampling frequency criteria, I'd like to do more confirmation about safe detection.
I also appreciate your remark of bug fixes, I'll revise them separately.

IDC-Dragon committed:
Squashed commit of the following:

commit d4c9df459c4a23e91d9bba3e617f3a2036b54491
Author: Jörg Hohensohn <joerg.hohensohn@gmx.de>
Date:   Sun Mar 5 09:03:45 2023 +0100

    return to uint64_t for prev_time, have a safe way for foreground read

commit e18267b4440c159da5b54901f6288cb2959b44d7
Author: Jörg Hohensohn <joerg.hohensohn@gmx.de>
Date:   Thu Mar 2 08:49:56 2023 +0100

    fix: stable_mask was one bit shorter than intended, cosmetic: improved readability

commit a038858bbbaeb559864cafa35f6d7cac60e86bed
Author: Jörg Hohensohn <joerg.hohensohn@gmx.de>
Date:   Wed Mar 1 22:44:18 2023 +0100

    speed up the ISR by removing float arithmetic, array loop, search loop and validation, that's now done in the foreground functions

IDC-Dragon commented:
Oh, this just crossed. To close the small foreground hole, I reverted to 64 bit time storage, and added a safe access function. Squashed the 3 commits together, for no clutter.
